sql/stmtdiagnostics: TestDiagnosticsRequest failed

sql/stmtdiagnostics.TestDiagnosticsRequest [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15802667?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15802667?buildTab=artifacts#/) on master @ [4848b258a1b04437b40e4cfce2f3a1384fdde435](https://github.com/cockroachdb/cockroach/commits/4848b258a1b04437b40e4cfce2f3a1384fdde435):


```
=== RUN   TestDiagnosticsRequest
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/85a915e05ca8994a6ad8cbe3d0421959/logTestDiagnosticsRequest3176496250
    test_log_scope.go:81: use -show-logs to present logs inline
    statement_diagnostics_test.go:67: error scanning '&{0x4004adf440 <nil>}': pq: query execution canceled due to statement timeout
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/85a915e05ca8994a6ad8cbe3d0421959/logTestDiagnosticsRequest3176496250
--- FAIL: TestDiagnosticsRequest (3.26s)
=== RUN   TestDiagnosticsRequest/timeout
    testing.go:1576: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
    --- FAIL: TestDiagnosticsRequest/timeout (0.25s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDiagnosticsRequest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39779

--- a/pkg/sql/stmtdiagnostics/statement_diagnostics_test.go
+++ b/pkg/sql/stmtdiagnostics/statement_diagnostics_test.go
@@ -163,15 +163,16 @@ func TestDiagnosticsRequest(t *testing.T) {
 		require.NoError(t, err)
 		checkNotCompleted(reqID)
 
-		// Set the statement timeout (as well as clean it up in a defer).
+		// Set the statement timeout.
 		runner.Exec(t, "SET statement_timeout = '100ms';")
-		defer func() {
-			runner.Exec(t, "RESET statement_timeout;")
-		}()
 
 		// Run the query that times out.
 		_, err = db.Exec("SELECT pg_sleep(999999)")
 		require.True(t, strings.Contains(err.Error(), sqlerrors.QueryTimeoutError.Error()))
+
+		// Reset the stmt timeout so that it doesn't affect the query in
+		// checkCompleted.
+		runner.Exec(t, "RESET statement_timeout;")
 		checkCompleted(reqID)
 	})
 

